src/libmongoc/tests/unified/operation.c

```diff
@@ -141,19 +141,36 @@ operation_list_databases (test_t *test,
    bool ret = false;
    mongoc_client_t *client = NULL;
    mongoc_cursor_t *cursor = NULL;
+   bson_parser_t *parser = NULL;
+   bson_t *opts = NULL;
+
+   parser = bson_parser_new ();
+   bson_parser_allow_extra (parser, true);
+
+   opts = bson_copy (bson_parser_get_extra (parser));
+   if (op->session) {
+      if (!mongoc_client_session_append (op->session, opts, error)) {
+         goto done;
+      }
+   }
+   if (op->arguments) {
+      bson_concat (opts, op->arguments);
```

Comment:
Note: I think it makes sense for libmongoc to have a macro here to assert that `arguments` does not contain any unexpected keys before we end up passing it along to the API method's `opts` parameter (or concatenating it in this case).

As-is, it's quite possible that a spec test might introduce a new option that gets silently ignored by libmongoc (because it has yet to implement the functionality) and the test might still pass.

I do this in PHPLIB's test runner in a few places, but I don't believe it's done for method arguments. Going to make a ticket to track that todo item, but I'd suggest creating one for CDRIVER as well if @kevinAlbs agrees.

Reply:
Extra keys can be checked with the `bson_parser_t`. By default will error on extra keys unless `bson_parser_allow_extra` is called. Some tests, but not all, already check the arguments and error on extra keys. Example from `operation_bulk_write`:

```
   parser = bson_parser_new ();
   bson_parser_array (parser, "requests", &requests);
   bson_parser_bool_optional (parser, "ordered", &ordered);
   if (!bson_parser_parse (parser, op->arguments, error)) {
      goto done;
   }
```

I agree it is preferable to error on unrecognized operation arguments. I think a minor follow-up ticket to make all cases explicit seems reasonable. But it seems low risk to me. If we do ignore an argument, a test should ideally fail.

Reply:
Filed a follow-up ticket to ensure explicit cases for all arguments to all operations as in the example above (CDRIVER-4018).